samples/nrf9160/nrf_cloud_rest_api/src/main.c

```diff
@@ -0,0 +1,150 @@
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <net/socket.h>
+#include <net/net_ip.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/at_cmd.h>
+#include <modem/at_notif.h>
+#include <net/tls_credentials.h>
+#include <modem/lte_lc.h>
+#include <modem/modem_key_mgmt.h>
+#include <net/http_client.h>
+#include <sys/base64.h>
+#include <net/nrf_cloud_rest.h>
+#include <modem/modem_jwt.h>
+#include <logging/log.h>
+
+BUILD_ASSERT(sizeof(CONFIG_REST_DEVICE_ID) > 1, "Device ID must be specified");
+
+LOG_MODULE_REGISTER(rest_sample, CONFIG_NRF_CLOUD_REST_SAMPLE_LOG_LEVEL);
+
+int init_modem_and_connect(void)
+{
+	int err;
+
+	err = nrf_modem_lib_init(NORMAL_MODE);
+	if (err) {
+		LOG_ERR("Failed to initialize modem library: %d", err);
+		return err;
+	}
+
+	err = at_cmd_init();
+	if (err) {
+		LOG_ERR("Failed to initialize AT commands, err %d", err);
+		return err;
+	}
+
+	err = at_notif_init();
+	if (err) {
+		LOG_ERR("Failed to initialize AT notifications, err %d", err);
+		return err;
+	}
+
+	LOG_INF("Waiting for network...");
+	err = lte_lc_init_and_connect();
+	if (err) {
+		LOG_ERR("Failed to connect to the LTE network, err %d", err);
+		return err;
+	}
+	LOG_INF("Connected");
+	return 0;
+}
+
+void main(void)
+{
+	char rx_buf[4096];
+	static struct nrf_cloud_fota_job_info job;
+	struct cell_based_loc_data location;
+	struct jwt_data jwt = { .subject = CONFIG_REST_DEVICE_ID,
+				.audience = "nRF Cloud",
+				.exp_delta_s = 0,
+				.sec_tag = CONFIG_REST_JWT_SEC_TAG,
+				.key_type = 2
+	};
+	struct nrf_cloud_rest_context rest_ctx = { .connect_socket = -1,
+						   .keep_alive = true,
+						   .auth = CONFIG_REST_API_TOKEN,
+						   .timeout_ms = 30000,
+						   .rx_buf = rx_buf,
+						   .rx_buf_len = sizeof(rx_buf)
+	};
+	struct nrf_cloud_rest_single_cell_request req = { .device_id = CONFIG_REST_DEVICE_ID,
+							  .mcc = 310,
+							  .mnc = 410,
+							  .area_code = 36879,
+							  .cell_id = 84486415};
+
+	int err = init_modem_and_connect();
+
+	if (err){
+		return;
+	}
+
+	/* Use API token for SCELL endpoint */
+	err = nrf_cloud_rest_get_single_cell_loc(&rest_ctx, &req, &location);
+	if (err) {
+		LOG_ERR("Single Cell API call failed, error: %d", err);
+	} else {
+		LOG_INF("Single Cell Response: %s", rest_ctx.response);
+	}
+
+	/* Use JWT for FOTA endpoints */
+	err = modem_jwt_generate(&jwt);
+	if (err) {
+		LOG_ERR("Failed to generate JWT, err %d", err);
+		return;
+	}
+
+	LOG_INF("JWT:\n%s",jwt.jwt_out);
+
+	/* JWT auth */
+	rest_ctx.auth = jwt.jwt_out;
+
+	err = nrf_cloud_rest_get_fota_job(&rest_ctx, CONFIG_REST_DEVICE_ID, &job);
+	if (err) {
+		goto cleanup;
+	}
+
+	if (job.type == NRF_CLOUD_FOTA_TYPE__INVALID) {
+		LOG_INF("No pending FOTA jobs");
+		goto cleanup;
+	}
+
+	LOG_INF("FOTA Job: %s, type: %d", log_strdup(job.id), job.type);
+
+	/* Disconnect after next API call */
+	rest_ctx.keep_alive = false;
+
+	/* Mark the job as succeeded */
```

Comment:
Does this actually do the FOTA update somewhere else, or is that left as an exercise for the reader?

Reply:
Yeah this is just to exercise the api.

Reply:
You might leave a comment to that effect, in case someone starts with this to develop something real.